File: program/program/implement-binary-search/implement-binary-search.jl
function binary_search(arr, x)
    lower_bound = 1
    upper_bound = length(arr)
    
    while lower_bound <= upper_bound
        mid_point = lower_bound + (upper_bound - lower_bound) ÷ 2
        
        if arr[mid_point] < x
            lower_bound = mid_point + 1
        elseif arr[mid_point] > x
            upper_bound = mid_point - 1
        else
            return mid_point  # x found at mid_point
        end
    end
    
    return -1  # x does not exist
end

# Example usage
arr = [2, 3, 4, 10, 40]
x = 10

result = binary_search(arr, x)
if result != -1
    println("Element $x is present at index $result")
else
    println("Element $x is not present in array")
end

File: program/program/implement-binary-search/ImplementBinarySearch.fs
let rec binarySearch (arr: 'a[]) (x: 'a) (lowerBound: int) (upperBound: int) : int =
    if upperBound < lowerBound then
        -1 // x does not exist
    else
        let midPoint = lowerBound + (upperBound - lowerBound) / 2
        if arr.[midPoint] < x then
            binarySearch arr x (midPoint + 1) upperBound
        elif arr.[midPoint] > x then
            binarySearch arr x lowerBound (midPoint - 1)
        else
            midPoint // x found at midPoint

// Example usage
let arr = [|2; 3; 4; 10; 40|]
let x = 10
let result = binarySearch arr x 0 (Array.length arr - 1)
if result <> -1 then
    printfn "Element %d is present at index %d" x result
else
    printfn "Element %d is not present in array" x
